core/src/main/java/com/amplifyframework/core/store/EncryptedKeyValueRepository.kt

```diff
@@ -17,93 +17,193 @@ package com.amplifyframework.core.store
+        } else {
+            // This repository was encrypted with the default master key
+            openKeystoreWithDefaultMasterKey(identifier)
+        }
+    }
+
+    private fun openKeystoreWithAmplifyMasterKey(identifier: String): SharedPreferences {
+        var amplifyMasterKey = getMasterKeyOrNull(amplifyMasterKeySpec)
+        if (amplifyMasterKey == null) {
+            logger.warn("Unable to retrieve Amplify master key. Deleting invalid master key and creating new one")
+            deleteMasterKey(amplifyMasterKeySpec)
+            amplifyMasterKey = getMasterKeyOrThrow(amplifyMasterKeySpec)
+        }
+
+        val fileName = getSharedPrefsFileName(identifier)
+
+        // Return the shared preferences if we can
+        getSharedPreferencesOrNull(fileName, amplifyMasterKey)?.let { return it }
+
+        logger.warn("Cannot retrieve preferences encrypted with amplify master key. Deleting and recreating.")
+        deleteSharedPreferences(fileName)
+        return getSharedPreferencesOrThrow(fileName, amplifyMasterKey)
+    }
+
+    private fun openKeystoreWithDefaultMasterKey(identifier: String): SharedPreferences {
+        // Try to open the encrypted preferences using the default master key
+        getMasterKeyOrNull(defaultMasterKeySpec)?.let { defaultMasterKey ->
+            val fileName = getSharedPrefsFileName(identifier)
+            getSharedPreferencesOrNull(fileName, defaultMasterKey)?.let { return it }
+        }
+
+        logger.warn("Cannot retrieve preferences encrypted with default master key. Deleting and recreating.")
+        // Delete the existing shared preferences file
+        deleteSharedPreferences(getSharedPrefsFileName(identifier))
+        // Create a new identifier with the amplify prefix
+        val newIdentifier = createInstallationIdentifier(getInstallationFile())
+        // Use the amplify master key to create the new shared preferences
+        return openKeystoreWithAmplifyMasterKey(newIdentifier)
+    }
+
+    private fun getSharedPreferencesOrNull(fileName: String, key: String) = try {
+        getSharedPreferencesOrThrow(fileName = fileName, key = key)
+    } catch (e: Exception) { null }
+
+    private fun getSharedPreferencesOrThrow(fileName: String, key: String): SharedPreferences =
         EncryptedSharedPreferences.create(
-            "$sharedPreferencesName.${getInstallationIdentifier(context, sharedPreferencesName)}",
-            MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC),
+            fileName,
+            key,
             context,
             AES256_SIV,
             AES256_GCM
         )
-    }
 
-    @VisibleForTesting
-    internal val editor: SharedPreferences.Editor by lazy {
-        sharedPreferences.edit()
+    private fun deleteSharedPreferences(fileName: String) = context.deleteSharedPreferences(fileName)
+
+    private fun deleteMasterKey(spec: KeyGenParameterSpec) = KeyStore.getInstance("AndroidKeyStore").run {
```

Comment:
should we make this method more focused? Ex: deleteAmplifyMasterKey? Especially since we would never want to accidentally introduce deleting app master key.

Reply:
Good callout, initially I was working with multiple master keys before settling on just one. I'll update.